codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/OperationMetricsStruct.java

```diff
@@ -0,0 +1,187 @@
+                    if err != nil {
+                        return nil, err
+                    }
+
+                    return context.WithValue(parent, operationMetricsKey{}, om), nil
+                }
+
+                func operationMetricTimer(m metrics.Meter, name, desc string) (metrics.Float64Histogram, error) {
+                    return m.Float64Histogram(name, func(o *metrics.InstrumentOptions) {
+                        o.UnitLabel = "s"
+                        o.Description = desc
+                    })
+                }
+
+                func getOperationMetrics(ctx context.Context) *operationMetrics {
+                    return ctx.Value(operationMetricsKey{}).(*operationMetrics)
+                }
+                """, scope);
+    }
+
+    private GoWriter.Writable generateHelpers() {
+        return goTemplate("""
+                func (m *operationMetrics) histogramFor(name string) metrics.Float64Histogram {
+                    switch name {
+                    case "client.call.duration":
+                        return m.Duration
+                    case "client.call.serialization_duration":
+                        return m.SerializeDuration
+                    case "client.call.resolve_identity_duration":
+                        return m.GetIdentityDuration
+                    case "client.call.resolve_endpoint_duration":
+                        return m.ResolveEndpointDuration
+                    case "client.call.signing_duration":
+                        return m.SignRequestDuration
+                    case "client.call.deserialization_duration":
+                        return m.DeserializeDuration
+                    default:
+                        panic("unrecognized operation metric")
+                    }
+                }
+
+                func timeOperationMetric[T any](
+                    ctx context.Context, metric string, fn func() (T, error),
+                    opts ...metrics.RecordMetricOption,
+                ) (T, error) {
+                    instr := getOperationMetrics(ctx).histogramFor(metric)
+                    opts = append([]metrics.RecordMetricOption{withOperationMetadata(ctx)}, opts...)
+
+                    start := time.Now()
+                    v, err := fn()
+                    end := time.Now()
+
+                    elapsed := end.Sub(start)
+                    instr.Record(ctx, float64(elapsed)/1e9, opts...)
+                    return v, err
+                }
+
+                func startMetricTimer(ctx context.Context, metric string, opts ...metrics.RecordMetricOption) func() {
+                    instr := getOperationMetrics(ctx).histogramFor(metric)
+                    opts = append([]metrics.RecordMetricOption{withOperationMetadata(ctx)}, opts...)
+
+                    var ended bool
+                    start := time.Now()
+                    return func() {
+                        if ended {
```

Comment:
No race conditions, since it's just in the context of a single operation.

The reason this is here is just to make the "stop" call idempotent -- that way you can just have a defer() on it which will catch all of the error cases in the middleware, and a direct call to it when the timed operation actually ends.